### DISCUSSION TITLE
indexes duplication on conflict do update

### DISCUSSION THREAD

**Discussion by zhuravlstrogo** [sqlalchemy/sqlalchemy#8607] [Usage Questions] [answered]
I create table:
```
create table allocations(
    allocation_id int not null generated always as identity,

   …

    constraint allocations_unique unique(
        project_id,
        resource_id,
        allocation_date,
        status
    ),
    primary key(allocation_id)
);
```

allocation_id is used instead of a composite primary key (project_id, resource_id, allocation_date, "status").

I update data with sqlalchemy:
```
    stmt = insert(Allocations).values(some_values)
    stmt = stmt.on_conflict_do_update(
            constraint='allocations_unique',
            set_={
                "project_id" : stmt.excluded.project_id,
                "resource_id" : stmt.excluded.resource_id,
                "allocation_date" : stmt.excluded.allocation_date,
                "status" : stmt.excluded.status,
        …
            })

    with engine.begin() as conn:
        conn.execute(stmt)
```

With each update of the table the indexes in the database are duplicated [screen](https://drive.google.com/file/d/1LuGvkR65jF4200QkvOQvFRFEBt0D8Qr_/view?usp=sharing)

Is it possible to avoid duplication of indexes?

**Comment by zzzeek** [comment, marked as answer]
hi -

running an INSERT statement won't create new indexes in the database.   you would have to be running some kind of CREATE INDEX statement somewhere, which is not seen here.    check that you don't have some kind of create_all() or other kind of migration that's linked to your operation that is running the statement.

**Reply by zhuravlstrogo** [reply]
I checked this before asking. Maybe the problem is in this insert `from sqlalchemy.dialects.postgresql import insert`. Also there is no problem when there is no index or unique constraint in the table

**Reply by zzzeek** [reply]
I can't provide more advice without a runnable example that reproduces the problem.

**Reply by zhuravlstrogo** [reply]
I found my mistake. 
Each time the table is updated, I pass the class Allocations in insert() with unique constraints. So constraints are created every time.
```
class Allocations(Base):
    __tablename__ = "allocations"
    __table_args__ = (
        UniqueConstraint('project_id', 'resource_id', 'allocation_date', 'status', name='allocations_unique'),
        {"schema": TEMPUS_STG_SCHEMA})
                      
    allocation_id = Column(Integer, primary_key=True)
    project_id = Column(Integer)
    resource_id = Column(Integer)
    status = Column(Integer)
    allocation_date = Column(DateTime)
  ```  
 Sorry for disturbing!
